src/relay_pic/main.c

```diff
@@ -0,0 +1,153 @@
+    PEIE = 1;
+    SSP1IF = 0;
+    SSP1IE = 1;
+}
+
+void interrupt i2cSlaveInt(void) { 
+    if (SSP1IF == 1) {
+        uint16_t temp;
+        SSPCONbits.CKP = 0;
+       
+        // If overflow or collision.
+        if ((SSPCONbits.SSPOV) || (SSPCONbits.WCOL)) {
+            temp = SSPBUF; // Read the previous value to clear the buffer
+            SSPCONbits.SSPOV = 0; // Clear the overflow flag
+            SSPCONbits.WCOL = 0;   // Clear the collision bit
+            SSPCONbits.CKP = 1;
+        }
+
+        // If last byte was Address + write
+        if (!SSPSTATbits.D_nA && !SSPSTATbits.R_nW) {
+            temp = SSPBUF;
+            while(!BF);
+            i2cSlaveRecv = SSPBUF;
+            SSPCONbits.CKP = 1;
+            SSPM3 = 0;
+       }
+        
+       // If last byte was Address + read
+       else if(!SSPSTATbits.D_nA && SSPSTATbits.R_nW) {
+           temp = SSPBUF;
+           BF = 0;
+           SSPBUF = i2cSlaveSend;
+           SSPCONbits.CKP = 1;
+           while(SSPSTATbits.BF);
+       }
+       
+       SSP1IF = 0;
+    } // if
+}
+
+//void readAnalogInputs() {
+//    for (int i = 0; i < 1000; i++) {}
+//    ADCON0 = 0x01; // Turn ADC on
+//    ADCON0 |= 1 << 1; // set b[1] "go" bit
+//    uint8_t doneBit;
+//    do { //wait for ADC to complete (go bit switches to 0 automatically when done)
+//        doneBit = ADCON0 & (1 << 1);
+//    } while (doneBit); //while go bit is on (AD conversion in progress)
+//
+//    adcResult = (ADRESH << 8) | ADRESL; //combine two 8bit values into a 16bit value
+//
+//    ADCON0 = 0x00; //Turn ADC off return;
+//}
+
+void setPower(bool out) { 
+    if (out) { LATA |= POWER;} 
+    else { LATA &= ~POWER; }
+}
+
+void setSelect(bool out) {
+    if (out) { LATA |= SELECT;} 
+    else { LATA &= ~SELECT; }
+}
+
+void setLim1(bool out) {
+    if (out) { LATA |= LIM1;} 
+    else { LATA &= ~LIM1; }
+}
+
+void setLim2(bool out) {
+    if (out) { LATA |= LIM2; } 
+    else { LATA &= ~LIM2; }
+}
+
+void setLed(bool out) {
+    if (out) { LATB |= LED; }
+    else { LATB &= ~LED; }
+}
+
+void readDipInputs() {
+    uint16_t newDip = 0;
+    newDip |= (PORTA & DIP1) ? (1) : 0;
+    newDip |= (PORTB & DIP2) ? (1 << 1) : 0;
+    newDip |= (PORTB & DIP3) ? (1 << 2) : 0;
+    newDip |= (PORTB & DIP4) ? (1 << 3) : 0;
+    dipInputs = newDip;
+}
+
+void setup() {
+    TRISA = CURR_SENSE1 | CURR_SENSE2 | DIP1;
+    TRISB = DIP2 | DIP3 | DIP4;
+    ANSELA = CURR_SENSE1 | CURR_SENSE2; //Set RA0 and RA1 as analog inputs.
+
+    // Configure ADC module 
+    // b[7] sets right justification, b[6:4] sets CS = FRC,
+    // b[2]+b[1:0] sets Vss and Vdd as references.
+
+    ADCON1 = 0b11110000;
+}
+
+int main(int argc, char** argv) {
+    setup();
+    setPower(true);
+    setSelect(true);
+    setLim1(true);
+    setLim2(true);
+    readDipInputs();
+    i2cSlaveInit(dipInputs);
+    while (1) {
+        // Infinite Loop
+        setLed(1);
+        readDipInputs();
+        //readAnalogInputs();

```

Comment:
General note: MCC allows you do define a struct of the channels you're interested in, giving them readable names. For example, in the injector code we have
https://github.com/waterloo-rocketry/cansw_injector/blob/aa1d5e417303065f9a181f73eedb278ddf167348/mcc_generated_files/adcc.h#L89
and
https://github.com/waterloo-rocketry/cansw_injector/blob/aa1d5e417303065f9a181f73eedb278ddf167348/error_checks.c#L18

Reply:
You can do a similar thing with digital pins if you're so inclined.

Reply:
Once again this keeps you from mucking around with things like `(1 << CURR_SENSE)`.